Add rpc.functions.async_execution decorator for TorchScript functions

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39505 [Debug-Only] Explicitly decref in UnpickledPythonCall dtor
* **#39267 Add rpc.functions.async_execution decorator for TorchScript functions**

When combined with `@torch.jit.script`, the order of decorators matter.
`@rpc.functions.async_execution` must be the outmost one. The
`async_execution` decorator will store the TorchScript function in
attribute `_wrapped_async_rpc_function` on the wrapper function, and
pass this wrapped TorchScript function (i.e., an instance of
`torch.jit.ScriptFunction`) to RPC. The caller will mark the ScriptCall
with `isAsyncExecution=true`, and the callee will extract the returned
`Future` in C++ and install subsequent processing as a callback to
that `Future`.

Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

mrshenli committed:
Add rpc.async_function decorator for TorchScript functions

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"




[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"




[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"


Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"


Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"


Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"


Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

pritamdamania87 commented:
Can we add a detailed summary for this PR too?

pritamdamania87 commented on torch/distributed/rpc/api.py:
```diff
@@ -527,11 +527,21 @@ def _invoke_rpc(to, func, rpc_type, args=None, kwargs=None, rpc_timeout=UNSET_RP
 
         is_async_fn = hasattr(func, "_wrapped_async_rpc_function")
 
+        if is_async_fn:
+            wrapped = func._wrapped_async_rpc_function
+            if isinstance(wrapped, torch.jit.ScriptFunction):
+                func = wrapped
```
This wouldn't work if we reverse the order of the decorators (jit.script first and then async_function)? I'm not sure if its possible to make this work such that the user doesn't have to depend on an order (ex: not sure if we can script a method that already has a decorator).

Hopefully we can make this order agnostic and add an appropriate unit test verifying this. If not, can we add this limitation to the documentation of `async_function`?

mrshenli replied:
Good point, will me test and document.

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"


Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"


Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"


Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"


Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"


Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"


Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"




Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"




Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"




Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"




Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"




Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for TorchScript functions"




Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for TorchScript functions"


When combined with `@torch.jit.script`, the order of decorators matter.
`@rpc.functions.async_execution` must be the outmost one. The
`async_execution` decorator will store the TorchScript function in
attribute `_wrapped_async_rpc_function` on the wrapper function, and
pass this wrapped TorchScript function (i.e., an instance of
`torch.jit.ScriptFunction`) to RPC. The caller will mark the ScriptCall
with `isAsyncExecution=true`, and the callee will extract the returned
`Future` in C++ and install subsequent processing as a callback to
that `Future`.

Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for TorchScript functions"


When combined with `@torch.jit.script`, the order of decorators matter.
`@rpc.functions.async_execution` must be the outmost one. The
`async_execution` decorator will store the TorchScript function in
attribute `_wrapped_async_rpc_function` on the wrapper function, and
pass this wrapped TorchScript function (i.e., an instance of
`torch.jit.ScriptFunction`) to RPC. The caller will mark the ScriptCall
with `isAsyncExecution=true`, and the callee will extract the returned
`Future` in C++ and install subsequent processing as a callback to
that `Future`.

Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for TorchScript functions"


When combined with `@torch.jit.script`, the order of decorators matter.
`@rpc.functions.async_execution` must be the outmost one. The
`async_execution` decorator will store the TorchScript function in
attribute `_wrapped_async_rpc_function` on the wrapper function, and
pass this wrapped TorchScript function (i.e., an instance of
`torch.jit.ScriptFunction`) to RPC. The caller will mark the ScriptCall
with `isAsyncExecution=true`, and the callee will extract the returned
`Future` in C++ and install subsequent processing as a callback to
that `Future`.

Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -1084,4 +1084,13 @@ def test_async_function_wrong_return_type(self):
+
+    @dist_init
+    def test_async_function_wrong_decorator_order(self):
+        with self.assertRaises(RuntimeError):
```
What error is raised here? Is this something that `torch.jit.script` complains about since there is another decorator?

mrshenli replied:
`torch.jit.script` failed to compile this function. It says it does not recognize `rpc`. Let me add the error to the comments. 

The reason for not checking the error string is to avoid making JIT error handling depending on RPC tests.

pritamdamania87 commented on torch/distributed/rpc/functions.py:
```diff
@@ -49,6 +49,32 @@ def async_execution(fn):
         >>> )
         >>> print(ret)  # prints tensor([3., 3.])
+
+        When combined with TorchScript functions, this decorator must be the
+        outmost one.
```
nit: should we also mention that if we're using any other decorators, this one should still be the outermost one.

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for TorchScript functions"


When combined with `@torch.jit.script`, the order of decorators matter.
`@rpc.functions.async_execution` must be the outmost one. The
`async_execution` decorator will store the TorchScript function in
attribute `_wrapped_async_rpc_function` on the wrapper function, and
pass this wrapped TorchScript function (i.e., an instance of
`torch.jit.ScriptFunction`) to RPC. The caller will mark the ScriptCall
with `isAsyncExecution=true`, and the callee will extract the returned
`Future` in C++ and install subsequent processing as a callback to
that `Future`.

Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for TorchScript functions"


When combined with `@torch.jit.script`, the order of decorators matter.
`@rpc.functions.async_execution` must be the outmost one. The
`async_execution` decorator will store the TorchScript function in
attribute `_wrapped_async_rpc_function` on the wrapper function, and
pass this wrapped TorchScript function (i.e., an instance of
`torch.jit.ScriptFunction`) to RPC. The caller will mark the ScriptCall
with `isAsyncExecution=true`, and the callee will extract the returned
`Future` in C++ and install subsequent processing as a callback to
that `Future`.

Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for TorchScript functions"


When combined with `@torch.jit.script`, the order of decorators matter.
`@rpc.functions.async_execution` must be the outmost one. The
`async_execution` decorator will store the TorchScript function in
attribute `_wrapped_async_rpc_function` on the wrapper function, and
pass this wrapped TorchScript function (i.e., an instance of
`torch.jit.ScriptFunction`) to RPC. The caller will mark the ScriptCall
with `isAsyncExecution=true`, and the callee will extract the returned
`Future` in C++ and install subsequent processing as a callback to
that `Future`.

Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for TorchScript functions"


When combined with `@torch.jit.script`, the order of decorators matter.
`@rpc.functions.async_execution` must be the outmost one. The
`async_execution` decorator will store the TorchScript function in
attribute `_wrapped_async_rpc_function` on the wrapper function, and
pass this wrapped TorchScript function (i.e., an instance of
`torch.jit.ScriptFunction`) to RPC. The caller will mark the ScriptCall
with `isAsyncExecution=true`, and the callee will extract the returned
`Future` in C++ and install subsequent processing as a callback to
that `Future`.

Differential Revision: [D21792688](https://our.internmc.facebook.com/intern/diff/D21792688)

[ghstack-poisoned]
